### PR TITLE
Generalize interface of InfluxDBPublisher to support more use cases (test-utils)

### PR DESCRIPTION
**Pull request by mosche**
Generalize interface of InfluxDBPublisher in `test-utils` to support more use cases (in preparation for #22238).

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [x] Update `CHANGES.md` with noteworthy changes.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by mosche** [comment]
R: @lgajowy 
R: @kkucharc

**Comment by mosche** [comment]
@kileys

**Comment by lukecwik** [comment]
Any of the people that you tagged seems to make sense but if they don't respond I can take a look once I'm back from vacation in two weeks.

Also, I understand that we could customize this but there seem to be tools that already do this for JMH like [this gradle plugin](https://github.com/mulesoft-labs/jmh-influx-report). Should we adopt one of these existing solutions or have you excluded them for some reason?

**Comment by mosche** [comment]
Thanks for having a look @lukecwik! Have a great time off :)

I've had a look at that plugin before. The reason why I'd be a bit reluctant to use the plugin is the lack of configuration. It uses the simple name of a benchmark class as measurement name. We might risk collision that way, especially across modules if we keep using a single db. But of course, we could start using multiple dbs with a db per JMH module. 

Another small point is that all JMH params + the test name are treated as fields (instead of tags) making these just one series per benchmark class where, in fact, each combination is - at least logically - a separate series of data. Though, that's certainly more cosmetic and probably done to keep series cardinality low. Also, from a query perspective it won't be very important as data points will be very infrequent.

**Comment by mosche** [comment]
Also, wondering a bit ... shouldn't test-utils just use the InfluxDB java client for simplicity?

**Comment by aromanenko-dev** [comment]
Run Spark ValidatesRunner

**Comment by aromanenko-dev** [comment]
Run Spark StructuredStreaming ValidatesRunner

**Comment by aromanenko-dev** [comment]
Run Spark Runner Nexmark Tests

File: sdks/java/testing/nexmark/src/main/java/org/apache/beam/sdk/nexmark/Main.java [hunk 1]
```diff
@@ -208,38 +195,24 @@ private static InfluxDBSettings getInfluxSettings(final NexmarkOptions options)
...
-    schemaResults.put("runner", runner);
-    schemaResults.put("measurement", measurement);
-
-    // By default, InfluxDB treats all number values as floats. We need to add 'i' suffix to
-    // interpret the value as an integer.
-    schemaResults.put("runtimeMs", runtimeMs + "i");
-    schemaResults.put("numResults", schemaResults.get("numResults") + "i");
-
-    return schemaResults;
+  private static InfluxDBPublisher.DataPoint influxDBDataPoint(
```
Comment (aromanenko-dev) [inline sdks/java/testing/nexmark/src/main/java/org/apache/beam/sdk/nexmark/Main.java:206]: nit: `createInfluxDBDataPoint()`

File: sdks/java/testing/nexmark/src/main/java/org/apache/beam/sdk/nexmark/Main.java [hunk 2]
```diff
@@ -208,38 +195,24 @@ private static InfluxDBSettings getInfluxSettings(final NexmarkOptions options)
         .get();
   }
 
-  private static String produceMeasurement(
-      final NexmarkOptions options, Map.Entry<NexmarkConfiguration, NexmarkPerf> entry) {
+  private static String measurementName(final NexmarkOptions options, NexmarkConfiguration config) {
```
Comment (aromanenko-dev) [inline sdks/java/testing/nexmark/src/main/java/org/apache/beam/sdk/nexmark/Main.java:198]: nit: `generateMeasurementName()`

File: sdks/java/testing/test-utils/src/main/java/org/apache/beam/sdk/testutils/publishing/InfluxDBPublisher.java [hunk 3]
```diff
@@ -83,73 +138,48 @@ private static void publishWithCheck(
...
-      final InfluxDBSettings settings,
-      final Map<String, String> tags)
-      throws Exception {
-
-    final HttpClientBuilder builder = provideHttpBuilder(settings);
-    final HttpPost postRequest = providePOSTRequest(settings);
-    final StringBuilder metricBuilder = new StringBuilder();
-
+  @VisibleForTesting
+  static String nexmarkDataPoints(
```
Comment (aromanenko-dev) [inline sdks/java/testing/test-utils/src/main/java/org/apache/beam/sdk/testutils/publishing/InfluxDBPublisher.java:142]: Should we extract this method into specific Nexmark class?

File: sdks/java/testing/test-utils/src/test/java/org/apache/beam/sdk/testutils/publishing/InfluxDBPublisherTest.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+  public void testNexmarkDataPoints() {
+    Map<String, Object> measurement =
+        ImmutableMap.<String, Object>builder()
+            .put("measurement", "name")
+            .put("timestamp", 9999L)
+            .put("runtimeMs", "1000i")
+            .put("numResults", "10i")
+            .build();
+    List<Map<String, Object>> measurements =
+        ImmutableList.of(measurement, measurement, measurement);
```
Comment (aromanenko-dev) [inline sdks/java/testing/test-utils/src/test/java/org/apache/beam/sdk/testutils/publishing/InfluxDBPublisherTest.java:41]: nit: create a list of measurements with different values

**Review by aromanenko-dev — APPROVED** [review]
It LGTM except some minor notes.

@lukecwik Kind ping. Do you have something to add?

**Comment by mosche** [comment]
Thanks a lot @aromanenko-dev and @lukecwik, I'm on PTO for the remainder of this month. I'll address as soon as I'm back.

File: sdks/java/testing/test-utils/src/main/java/org/apache/beam/sdk/testutils/publishing/InfluxDBPublisher.java (hunk 3 above)
Comment (mosche) [reply sdks/java/testing/test-utils/src/main/java/org/apache/beam/sdk/testutils/publishing/InfluxDBPublisher.java:142]: This is part of the legacy interface and to be removed, I marked it as deprecated and added a comment

File: sdks/java/testing/test-utils/src/main/java/org/apache/beam/sdk/testutils/NamedTestResult.java [hunk 5]
```diff
@@ -85,4 +85,17 @@ public String getMetric() {
...
+    return ImmutableMap.of("test_id", testId, "metric", metric);
+  }
+
+  public Map<String, Number> fields() {
+    return ImmutableMap.of("value", value);
+  }
+
+  /** Convert this result to InfluxDB data point. */
+  public InfluxDBPublisher.DataPoint toInfluxDBDataPoint(String measurement) {
+    return InfluxDBPublisher.dataPoint(measurement, tags(), fields(), null);
```
[the remark this answers is not in the document]
Comment (mosche) [reply sdks/java/testing/test-utils/src/main/java/org/apache/beam/sdk/testutils/NamedTestResult.java:99]: At the moment `timestamp` of `NamedTestResult` is ignored when writing to InfluxDB. TBH I'm not sure why, maybe it's incompatible with InfluxDB? For now I'd prefer to keep that as is to not introduce any potential problems.

> timestamp: Time at which this result was sampled. Should be in a BigQuery supported timestamp format.

File: sdks/java/testing/test-utils/src/main/java/org/apache/beam/sdk/testutils/publishing/InfluxDBPublisher.java [hunk 6]
```diff
@@ -46,35 +56,80 @@
...
+    DataPoint() {}
+
+    public abstract @Pure String measurement();
+
+    public abstract @Pure Map<String, String> tags();
+
+    public abstract @Pure Map<String, Number> fields();
+
+    @Nullable
+    public abstract @Pure Long timestamp();
```
[the remark this answers is not in the document]
Comment (mosche) [reply sdks/java/testing/test-utils/src/main/java/org/apache/beam/sdk/testutils/publishing/InfluxDBPublisher.java:80]: I renamed it to timestampSecs and added a comment to explain. Also, see below

File: sdks/java/testing/test-utils/src/main/java/org/apache/beam/sdk/testutils/publishing/InfluxDBPublisher.java [hunk 7]
```diff
@@ -83,73 +138,48 @@ private static void publishWithCheck(
...
+      Map<String, ?> fields,
+      @Nullable Object timestamp) {
+    checkState(!fields.isEmpty(), "fields cannot be empty");
+    builder.append(measurement);
+    tags.forEach((k, v) -> builder.append(',').append(k).append('=').append(v));
+    builder.append(' ');
+    fields.forEach((k, v) -> builder.append(k).append('=').append(fieldValue(v)).append(','));
+    builder.setLength(builder.length() - 1); // skip last comma
+    if (timestamp != null) {
+      builder.append(' ').append(timestamp);
```
[the remark this answers is not in the document]
Comment (mosche) [reply sdks/java/testing/test-utils/src/main/java/org/apache/beam/sdk/testutils/publishing/InfluxDBPublisher.java:174]: Thanks for having a close look, @lukecwik!
Turns out current precision is in seconds. The only place were the actual timestamp of the test is used is Nexmark:
```
if (options.getExportSummaryToInfluxDB()) {
        final long timestamp = start.getMillis() / 1000; // seconds
        savePerfsToInfluxDB(options, actual, timestamp);
      }
```
 The publisher configures `precision` as follows to override the default precision of InfluxDB. Again, 
``` 
return new HttpPost(settings.host + "/write?db=" + settings.database + "&" + retentionPolicy + "&precision=s");
```

**Comment by mosche** [comment]
@aromanenko-dev I've addressed the comments and pushed some changes to unblock your PR. Please merge when tests are green.
Ticket is https://github.com/apache/beam/issues/22238 as mentioned in the description.

**Comment by aromanenko-dev** [comment]
@mosche Many thanks for make it merged!

**Pull request merged**